module/jobqueue/README.md

```diff
@@ -0,0 +1,63 @@
+
+Let's just call this uint64 value "Job Index" or index.
+
+So if we iterate through each index from low to high, and find each job by index, then we are able to iterate through each job.
+
+Therefore modules.Job interface abstracts it into a method: `AtIndex`.
+
+`AtIndex` method returns the job at any given index.
+
+Job consumer relies on the modules.Jobs to find jobs. However, modules.Jobs doesn't provide a way to notify as soon as a new job is available. So it's consumer's job to keep track of the values returned by module.Jobs's `Head` method and find jobs that are new.
+
+### Using Check method to notify job consumer for checking new jobs
+Job consumer provides the Check method for users to notify new jobs available.
+
+Once called, job consumer will iterate through each height with the AtIndex method. It stops when one of the following condition is true:
+1. no job was found at a index
+2. no more workers to work on them
+3. max search ahead is reached (will explain later)
+
+Check method is concurrent safe, meaning even if job consumer is notified concurrently about new jobs available, job consumer will check at most once to find new jobs.
+
+Whenever a worker finishes a job, job consumer will also call Check internally.
+
+### Storing job consuming progress in storage.ConsumerProgress
+Job consumer stores the last processed job index in storage.ConsumerProgress, so that on startup, job consumer can read the last processed job index from storage and compare with the last available job index from modules.Jobs's Head method to resume job processing.
+
+This ensures each job will be processed at least once.
+
+### Using Workers to work on each job
+
+When Job consumer finds new job, it uses the Worker interface to process each job. Each job is also an interface, so it's user's responsiblity to handle the conversion between job and the actual job data type. So that the modules.Job can return a job to converted back to the actual job data for worker to work on.
+
+In the scenario of processing finalized blocks, implementing symmetric functions like BlockToJob and JobToBlock are recommended for this conversion.
+
+In order to report job completion, the worker needs to call job consumer's NotifyJobIsDone method.
+
+## Pipeline Pattern
+Multiple jobqueues can be combined to form a pipeline. This is useful in the scenario that the first job queue will process each finalized block and create jobs to process data depending on the block, and having the second job queue to process each job created by the worker of the first job queue.
+
+For instance, verification node uses 2-jobqueue pipeline to find chunks from each block and create jobs if the block has chunks that it needs to verify, and the second job queue will allow verification node to verify each chunk with a max number of workers.
```

Comment:
```suggestion
For instance, verification node uses 2-jobqueue pipeline to find chunks from each block and create jobs if the block has chunks that it needs to verify, and the second job queue will allow verification node to verify each chunk with a max number of workers.

## Considerations

### Push vs Pull
The jobqueue architecture is optimized for "pull" style processes, where the consumer pulls jobs from a source when workers are available. All current implementations are using this style since it lends well to asynchronously processing jobs based on block heights.

Some use cases require "push" style jobs where there is a job producer that creates jobs, and a consumer that processes work from the producer. This is possible with the jobqueue, but requires the producer persist the jobs to a database, then implement the `Head` and `AtIndex` methods that allow accessing jobs by sequential `uint64` indexes.

In existing pull use cases, the sync engine is effectively the job producer.
```